### DISCUSSION TITLE
Notification System for Tauri

### DISCUSSION THREAD

**Discussion by louremipsum** [tauri-apps/tauri#6604] [Q&A] [answered]
Hey folks,
I wanted to build a notification system for my Tauri app which supplies notifications which can execute an action or trigger a Redux action upon interaction. I checked out the notification API provided by Tauri but it seems like it can only send notifications with no interactivity.
I check the crate on which the API is based(I think I read this somewhere) which is [notify_rust](https://docs.rs/notify-rust/latest/notify_rust/index.html#). Now the problem is, that this crate is mainly supported for XDG(Linux) only and many [methods ](https://docs.rs/notify-rust/latest/notify_rust/index.html#platform-differences) straight up give compile errors on MacOS and Windows.

I've searched more on the net for alternatives for an interactive notification like toast-rust(Windows 10 only) or mac-notification-sys(for MacOS) but writing 2 separate implementations for other target OS seems like a chore
It is written that I can use the _**target_os** toggle_ if I want to use those methods which fail to compile but honestly I don't even know where to start for writing a windows and Mac version.

Are there any alternatives which provide the same or approx. same features as notify_rust but for all 3 platforms OR point me in some direction on how I would go about writing a windows/mac implementation using the target_os toggle.

Thank you for reading this long question, hoping to hear from the community soon :D

**Comment by FabianLars** [comment, marked as answer]
> Are there any alternatives which provide the same or approx. same features as notify_rust but for all 3 platforms

Hmm, i don't know of any alternative and i would imagine that it'd be the ecosystem standard over notify_rust if it would exist.

About the target_os stuff, did you see the small example in their docs? https://docs.rs/notify-rust/latest/notify_rust/index.html#toggles
Here's a small more explicit example:
```rs
let mut notification = Notification::new();
#[cfg(any(target_os = "windows", target_os = "linux"))] // For functions that don't work on macos
notification.appname("");
#[cfg(target_os = "macos")] // For functions that only work on macos
notification.some_macos_specific_fn();

notification.show();
```
At least that's what the docs meant with the toggle.

----

I don't think there's a better macos notification library than mac-notification-sys (which is used by notifiy_rust) yet, and i don't think the alternatives to winrt-notification (which is used by notifiy-rust) are any better either 🤔 
In other words i think you're stuck with whatever notify_rust offers unless you can contribute to it directly.

**Reply by louremipsum** [reply]
> Hmm, i don't know of any alternative and i would imagine that it'd be the ecosystem standard over notify_rust if it would exist.

Even when I knew the answer to that question, I wanted to be ignorant that maybe I missed something and there is a better library which will solve all my issues lol

I checked the examples in their docs but couldn't make out how would i implement any feature aside from starting from `#[cfg(target_os = "macos")]` so your example does help me in that respect.
since there is no alternative, I'll just write the implementation myself, maybe contribute if I'm successful.

**Reply by maoyaru123** [reply]
Hello, I would like to know if you have achieved it. I have encountered the same issue and would like to see if there is a mature solution available.
